### PR TITLE
SI-6642 Adds iteratorFrom, keysIteratorFrom, and valuesIteratorFrom

### PR DESCRIPTION
**Pull request by JamesIry**
SI-6642 Adds iteratorFrom, keysIteratorFrom, and valuesIteratorFrom

Adds the ability to efficiently create an iterator that starts at a
given key or element of a sorted set or map. Similar work is done for
key and value only iterators on maps.

The heart of the work is in RedBlackTree. Most of the rest is pushing
the new api methods throughout the appropriate spots in the collection
API.

Along the way mutable.TreeSet is reimplemented in terms of
RedBlackTree to eliminate code duplication and improve efficiency.

review @adriaanm

### CODE REVIEW

**Comment by soc** [comment]
Using plural for the methods `keysIteratorFrom` and `valuesIteratorFrom` sounds wrong because `Iterator` already conveys that potentially multiple elements are involved. What's the reasoning behind it?

**Comment by JamesIry** [comment]
Because we already have "keysIterator" and "valuesIterator". I'm just adding "from" variants in parallel with "iterator" => "iteratorFrom"

**Comment by soc** [comment]
Ugh ... who did that naming? :-/

**Comment by JamesIry** [comment]
Open a ticket. We can debate the merits of a rename (with deprecated forwarders) there.

File: src/library/scala/Enumeration.scala [hunk 1]
```diff
@@ -254,7 +254,8 @@ abstract class Enumeration (initial: Int) extends Serializable {
     def contains(v: Value) = nnIds contains (v.id - bottomId)
     def + (value: Value) = new ValueSet(nnIds + (value.id - bottomId))
     def - (value: Value) = new ValueSet(nnIds - (value.id - bottomId))
-    def iterator = nnIds.iterator map (id => thisenum.apply(id + bottomId))
+    def iterator = nnIds.iterator map (id => thisenum.apply(bottomId + id))
```
Comment (adriaanm) [inline src/library/scala/Enumeration.scala:257]: I guess my original comment was mainly about removing the unnecessary `.apply`, but I'll take it as-is.

**Comment by adriaanm** [comment]
LGTM

future work:
- do we have a ticket that schedules the removal of AVLTree?
- the tests are nice and mostly self-explanatory, but some comments there would still be good

**Comment by JamesIry** [comment]
https://issues.scala-lang.org/browse/SI-7155

**Pull request merged**